proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,461 @@
+
+After this, you need to run your program with `dotnet run` and friends, but by virtue of being a project, you now have the full `dotnet` toolset at your disposal.
+
+`#!` directives wouldn't technically need to be removed. You should still be able to run a single file the same was as before. Having a project file in the current directory shouldn't change that.
+
+The second option has potentially wild implications across our entire build tools and IDE tooling stack. It would require us to fundamentally rethink how .NET applications can be built.
+
+For the purposes of this design, we will require that users create a project-based codebase if they wish to be grown up.
+
+## Proposed order of implementation
+
+A way to approach this is to break it up into phases. Only phase 1 has to be implemented first. The rest could come afterwards. Some could never be implemented at all. This is merely a proposed ordering.
+
+### Phase 1 - basic scenario only
+
+A first phase of implementation would be to support everything up until `#r "nuget:...` and coding against a framework/sdk. That would mean:
+
+* Core C# language changes (global `using`s and `#!` directive support)
+* Build tooling additions to support global `using`s
+* Build tooling additions to erect the right guardrails (e.g., not using `dotnet publish` here).
+* `dotnet add project` command support
+
+This would mean that you could write anything that you'd be able to write in today's console app project. You can also "graduate" to a project file to get access to the full range of `dotnet` commands, packages, etc.
+
+### Phase 2 - tooling support in VSCode
+
+One of the most popular tools for people learning C# and .NET, or just doing lightweight work, is VSCode. It would have to support Simple C# Programs with the full range of tooling support:
+
+* IntelliSense, navigation, etc.
+* Debugging (assuming you have the right `launch.json` set up)
+
+To support this, the C# experience in VSCode would have to know what to load/run at design-time to power these tools. Since there's no project file to evaluate, it would have to take a different approach.
+
+### Phase 2 - package and framework support
+
+This phase would bring in support for referencing packages and coding against a framewor/sdk:
+
+* `#r "nuget:..."` or similar directive that resolves packages. Also works in tooling, so you get IntelliSense for the package
+* `#!/usr/bin/share/dotnet-aspnet` or `#r "framework:..."` support, allowing you to run against ASP.NET Core (`server.cs`)
+
+This would require some design work to arrive at the best syntax and semantics. It is also expected that tooling works when these are incorporated.
+
+### Phase 3 - multiple files
+
+This phase would bring in support for pulling in multiple files for a single program. Imagine a `server.cs` with a `network-utils.cs` file, where `network-utils.cs` is used by `server.cs`.
+
+Some guardrails would have to be built into place.
```

Comment:
Considering if `#load "some other.cs"` should work like in fsx scripts, and identify if a single file has top level statements (that would be the program).